CMake/FindLDAP.cmake

```
@@ -0,0 +1,103 @@
+#***************************************************************************
+#                                  _   _ ____  _
+#  Project                     ___| | | |  _ \| |
+#                             / __| | | | |_) | |
+#                            | (__| |_| |  _ <| |___
+#                             \___|\___/|_| \_\_____|
+#
+# Copyright (C) Daniel Stenberg, <daniel@haxx.se>, et al.
+#
+# This software is licensed as described in the file COPYING, which
+# you should have received as part of this distribution. The terms
+# are also available at https://curl.se/docs/copyright.html.
+#
+# You may opt to use, copy, modify, merge, publish, distribute and/or sell
+# copies of the Software, and permit persons to whom the Software is
+# furnished to do so, under the terms of the COPYING file.
+#
+# This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+# KIND, either express or implied.
+#
+# SPDX-License-Identifier: curl
+#
+###########################################################################
+# Find the ldap library
+#
+# Input variables:
+#
+# - `LDAP_INCLUDE_DIR`:   The ldap include directory.
+# - `LDAP_LIBRARY`:       Path to `ldap` library.
+# - `LDAP_LBER_LIBRARY`:  Path to `lber` library.
+#
+# Result variables:
+#
+# - `LDAP_FOUND`:         System has ldap.
+# - `LDAP_INCLUDE_DIRS`:  The ldap include directories.
+# - `LDAP_LIBRARIES`:     The ldap library names.
+# - `LDAP_LIBRARY_DIRS`:  The ldap library directories.
+# - `LDAP_CFLAGS`:        Required compiler flags.
+# - `LDAP_VERSION`:       Version of ldap.
+
+if(CURL_USE_PKGCONFIG AND
+   NOT DEFINED LDAP_INCLUDE_DIR AND
+   NOT DEFINED LDAP_LIBRARY AND
+   NOT DEFINED LDAP_LBER_LIBRARY)
+  find_package(PkgConfig QUIET)
+  pkg_check_modules(LDAP "ldap")
+  pkg_check_modules(LDAP_LBER "lber")
+endif()
+
+if(LDAP_FOUND AND LDAP_LBER_FOUND)
+  list(APPEND LDAP_LIBRARIES ${LDAP_LBER_LIBRARIES})
+  list(REMOVE_DUPLICATES LDAP_LIBRARIES)
```

"Traditional" linkers rely on libary order. `REMOVE_DUPLICATES` would remove the late duplicates (e.g. the last `-lm`), and this breaks linking with "traditional" linkers which rely on lib order.

```suggestion
  list(REMOVE_ITEM LDAP_LIBRARIES ${LDAP_LBER_LIBRARIES})
  list(APPEND LDAP_LIBRARIES ${LDAP_LBER_LIBRARIES})
```

---

Note to self that this pattern is also used for mbedtls and brotli (pending PR).

And also for Libs.private in libcurl.pc.

I guess these can be dropped. Or come up with a function dropping the early dupes.

(There is also the "overlinking" issue when linking shared libs with these deps. Could't find a solution for that.)

---

Do you think this'd work?:
```
list(REVERSE LDAP_LIBRARIES)
list(REMOVE_DUPLICATES LDAP_LIBRARIES)
list(REVERSE LDAP_LIBRARIES)
```

---

I used to use this pattern, too.

In general, I just leave the duplicates untouched today (and patch out `REMOVE_DUPLICATES`). Different linkers may need different strategies (accoding to CMake). With Apple frameworks, you may accidentally remove `-framework`.